capa/features/extractors/binexport2/insn.py

```diff
@@ -0,0 +1,443 @@
+    rva = address - ctx.analysis.base_address
+
+    try:
+        if sample_bytes.startswith(capa.features.extractors.common.MATCH_PE):
+            pe = cache.get("pe")
+            if not pe:
+                pe = pefile.PE(data=sample_bytes)
+                cache["pe"] = pe
+            return pe.get_data(rva, size)
+        elif sample_bytes.startswith(capa.features.extractors.common.MATCH_ELF):
+            elf = cache.get("elf")
+            if not elf:
+                elf = ELFFile(io.BytesIO(sample_bytes))
+                cache["elf"] = elf
+
+            # ELF segments are for runtime data,
+            # ELF sections are for link-time data.
+            for segment in elf.iter_segments():
+                # assume p_align is consistent with addresses here.
+                # otherwise, should harden this loader.
+                segment_rva = segment.header.p_vaddr
+                segment_size = segment.header.p_memsz
+                if segment_rva <= rva < segment_rva + segment_size:
+                    segment_data = segment.data()
+
+                    # pad the section with NULLs
+                    # assume page alignment is already handled.
+                    # might need more hardening here.
+                    if len(segment_data) < segment_size:
+                        segment_data += b"\x00" * (segment_size - len(segment_data))
+
+                    segment_offset = rva - segment_rva
+                    return segment_data[segment_offset : segment_offset + size]
+
+            raise ReadMemoryError("address not mapped")
+        else:
+            logger.warning("unsupported format")
+            raise ReadMemoryError("unsupported file format")
+    except Exception as e:
+        raise ReadMemoryError("failed to read memory: " + str(e)) from e
+
+
+def extract_insn_bytes_features(fh: FunctionHandle, bbh: BBHandle, ih: InsnHandle) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    bbi: BasicBlockContext = bbh.inner
+    ii: InstructionContext = ih.inner
+
+    ctx = fhi.ctx
+    be2 = fhi.ctx.be2
+    sample_bytes = fhi.ctx.sample_bytes
+    idx = fhi.ctx.idx
+
+    basic_block_index = bbi.basic_block_index
+    instruction_index = ii.instruction_index
+
+    reference_addresses: List[int] = []
+
+    if instruction_index in idx.data_reference_index_by_source_instruction_index:
+        for data_reference_index in idx.data_reference_index_by_source_instruction_index[instruction_index]:
+            data_reference = be2.data_reference[data_reference_index]
+            data_reference_address = data_reference.address
+
+            reference_addresses.append(data_reference_address)
+
+    if (not reference_addresses) and _is_ghidra_symbol_madness(be2, instruction_index):
+        instruction = be2.instruction[ii.instruction_index]
+        mnemonic = be2.mnemonic[instruction.mnemonic_index]
+        mnemonic_name = mnemonic.name.lower()
+
+        if mnemonic_name == "adrp":
```

Comment:
we could add separate functions here for various architectures to split this up a bit nicer

Reply:
yeah, we definitely need to do something like that, once there are multiple supported archs. we've talked about doing the same for viv, too.